discord: add option to install Vencord

Scrumplex opened the pull request:
This PR moves `vencord-web-extension` from `webcord` to a top level package, allows building the injector scripts, and adds an option to `discord` to install Vencord.

I am unsure if `vencord` and `vencord-web-extension` should be top-level packages, as they are not really packages.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Scrumplex committed:
vencord-web-extension: init at 1.1.6

vencord-web-extension was previously a part of webcord (to build
webcord-vencord). As we want to use this for Discord as well, move it to
a top-level package.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
discord: add option to install Vencord

Add new withVencord parameter to install Vencord when building. This
uses the previously introduced vencord package.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex commented:
CC @Infinidoge @MP2E @Artturin for Discord

FlafyDev commented on pkgs/development/misc/vencord/default.nix:
```diff
@@ -48,13 +49,13 @@ buildNpmPackage rec {
   ];
 
   installPhase = ''
-    cp -r dist/extension-unpacked $out
```
maybe change it to:
`cp -r dist/ $out` for build
`cp -r dist/extension-unpacked $out` for buildWeb

This way only the necessary files for vencord web extension will be installed and there will be no need for the `/extension-unpacked` part in
`vencord = "${vencord-web-extension}/extension-unpacked";` (`webcord-vencord/default.nix`)

FlafyDev reviewed (changes requested):
Is `development/misc` the best directory for `vencord`?
I think `misc` would fit more, since I am not aware of any similar packages..

Infinidoge commented:
Has this been tested in combination with OpenASAR?

Scrumplex commented:
> Has this been tested in combination with OpenASAR?

It works just as well. Vencord will just load the entrypoint in the _app.asar file, no matter if it's OpenASAR or the original.

Infinidoge commented on pkgs/applications/networking/instant-messengers/discord/default.nix:
```diff
@@ -59,7 +59,7 @@ let
     sourceProvenance = with sourceTypes; [ binaryNativeCode ];
     license = licenses.unfree;
-    maintainers = with maintainers; [ MP2E artturin infinidoge jopejoe1 ];
+    maintainers = with maintainers; [ MP2E Scrumplex artturin infinidoge jopejoe1 ];
```
Nit: This should technically be in its own commit

Infinidoge commented on pkgs/misc/vencord/default.nix:
```diff
@@ -47,14 +48,16 @@ buildNpmPackage rec {
     homepage = "https://github.com/Vendicated/Vencord";
     license = licenses.gpl3Only;
-    maintainers = with maintainers; [ FlafyDev NotAShelf ];
+    maintainers = with maintainers; [ FlafyDev NotAShelf Scrumplex ];
```
Nit: This should be in its own commit

Infinidoge reviewed (changes requested):
Overall, great PR! Everything seems to be working on my machine.
Just a couple nitpicks about commits:
- To my knowledge, adding yourself as a maintainer to a package should be done as their own commits.
- It might be a good idea to separate the refactoring of webcord-vencord into its own commit, though that I'm unsure about. It would definitely be a good idea to explain in the commit message why things were moved around, though.

Scrumplex committed:
vencord: add Scrumplex to maintainers

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
discord: add Scrumplex to maintainers

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Scrumplex committed:
vencord: allow building Discord scripts

Vencord can be built in two different ways. Either as a web-extension,
or as a set of scripts to be injected into Discord Desktop.

Allow to select between these two modes using a parameter and expose
both variants as top-level packages.

This also renames vencord-web-extension to vencord.

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

Infinidoge reviewed (changes requested):
Tiniest of nitpicks: Since Vencord is becoming the top level package, it should probably be `vencord: init ...` rather than `vencord-web-extension: init ...`, since later commits will be `vencord: ... -> ...`

Scrumplex commented:
Yeah I wasn't sure about that part. Technically at that point I only add `vencord-web-extension` and later just refactor it to provide both `vencord` and `vencord-web-extension`

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2231

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/discord/linux.nix:
```diff
@@ -9,7 +9,7 @@
 , wayland
 , branch
-, common-updater-scripts, withOpenASAR ? false, withTTS ? false }:
+, common-updater-scripts, withOpenASAR ? false, withVencord ? false, withTTS ? false }:
```
```suggestion
, common-updater-scripts
, withOpenASAR ? false
, withVencord ? false, vencord
, withTTS ? false
}:
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/discord/linux.nix:
```diff
@@ -1,4 +1,4 @@
-{ pname, version, src, openasar, meta, binaryName, desktopName, autoPatchelfHook
+{ pname, version, src, openasar, vencord, meta, binaryName, desktopName, autoPatchelfHook
```
```suggestion
{ pname, version, src, openasar, meta, binaryName, desktopName, autoPatchelfHook
```

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -24195,6 +24195,10 @@ with pkgs;
 
+  vencord = callPackage ../misc/vencord { };
+
+  vencord-web-extension = callPackage ../misc/vencord { buildWebExtension = true; };
```
How about we move this to passthru?

Scrumplex replied:
Not quite sure. These are really two different build modes at the end.

`buildWebExtension = true;` will only build the web extension (for webcord), while `buildWebExtension = false;` will only build the variant for Discord

Nyabinary commented:
@Scrumplex Vencord is outdated I think
